File: src/runtime/java/mcp-mesh-sdk/src/main/java/io/mcpmesh/a2a/A2AStream.java
package io.mcpmesh.a2a;

import io.mcpmesh.JobController;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import tools.jackson.databind.JsonNode;
import tools.jackson.databind.ObjectMapper;
import tools.jackson.databind.node.ArrayNode;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.util.Iterator;
import java.util.NoSuchElementException;

/**
 * Iterator over parsed A2A SSE events — returned by
 * {@link A2AClient#subscribe}.
 *
 * <p>Implements {@link Iterable} so callers can write
 * {@code for (A2AEvent event : stream)} idiomatically. Implements
 * {@link AutoCloseable} so callers MUST either iterate to completion
 * (which auto-closes once the {@code isFinal=true} frame is consumed)
 * OR use try-with-resources to release the underlying connection
 * cleanly.
 *
 * <p>Mirrors {@code mesh._a2a_consumer.A2AStream} from the Python
 * runtime (issue #910 Phase 3) — uses plain JDK
 * {@link BufferedReader} instead of okio (no new Maven dependency).
 *
 * <h2>SSE parsing</h2>
 * Each event is preceded by zero or more {@code data:} lines and
 * terminated by an empty line. Lines starting with {@code :} (SSE
 * comment / keepalive) and any other field lines ({@code event:},
 * {@code id:}, {@code retry:}) are ignored. {@code data:} payloads
 * are concatenated with newlines per the SSE spec; the consumer
 * parses the joined buffer as JSON-RPC and translates known shapes
 * (status / artifact) into {@link A2AEvent}.
 *
 * <h2>Cancel propagation</h2>
 * Per A2A v1.0, client disconnect is a transient signal and does NOT
 * imply cancel — the producer continues running unless explicitly
 * canceled via {@code tasks/cancel}. {@link #bridge} therefore does
 * NOT POST {@code tasks/cancel} upstream on disconnect. Users who need
 * cancel propagation should use
 * {@link A2AClient#submit} + {@link A2AJob#bridge} instead, which
 * polls {@link JobController#isCancelled()} between iterations and
 * propagates upstream.
 *
 * <p><b>Threading:</b> SSE line parsing happens on the caller's thread.
 * Iteration via {@code for (A2AEvent event : stream)} blocks the calling
 * thread on each {@code next()} call while the next SSE frame is being
 * read from the network. {@link #bridge(JobController)} likewise runs
 * synchronously on the caller's thread until the stream terminates.
 */
public final class A2AStream implements Iterable<A2AEvent>, AutoCloseable {

    private static final Logger log = LoggerFactory.getLogger(A2AStream.class);

    private final HttpResponse<InputStream> response;
    private final BufferedReader reader;
    private final ObjectMapper objectMapper;
    private final String taskId;
    private volatile boolean closed;

    A2AStream(HttpResponse<InputStream> response, String taskId, ObjectMapper objectMapper) {
        this.response = response;
        this.objectMapper = objectMapper;
        this.taskId = taskId;
        // BufferedReader.readLine handles both \n and \r\n line
        // terminators, matches the SSE spec.
        this.reader = new BufferedReader(
            new InputStreamReader(response.body(), StandardCharsets.UTF_8));
    }

    /** The consumer-generated task ID echoed back by the producer. */
    public String taskId() {
        return taskId;
    }

    @Override
    public Iterator<A2AEvent> iterator() {
        return new SseIterator();
    }

    /**
     * Iterate events; mirror progress to the supplied
     * {@link JobController}; return the final artifact value when
     * the terminal frame arrives.
     *
     * <p>Returns the final artifact value: parsed JSON
     * ({@code Map<String,Object>} or {@code List<Object>}) when the
     * artifact text is valid JSON, otherwise the raw text
     * {@code String}. Empty artifacts return an empty {@code String}.
     *
     * <p>The framework's {@code task=true} wrapper handles
     * {@code controller.complete(...)} from the return; this method
     * only mirrors progress + propagates terminal state.
     *
     * @throws A2AJobFailedException   on terminal {@code state=failed}
     *                                 OR stream end without artifact.
     * @throws A2AJobCanceledException on terminal {@code state=canceled}.
     */
    public Object bridge(JobController controller) {
        if (controller == null) {
            throw new IllegalArgumentException("A2AStream.bridge: controller must be non-null");
        }
        return bridgeInternal(JobControllerAdapter.wrap(controller));
    }

    /**
     * Package-private bridge entry point that takes the test seam
     * adapter directly. Only {@link #bridge(JobController)} (production
     * callers) and the unit test in {@code A2AStreamTest} should use
     * this.
     */
    Object bridgeInternal(JobControllerAdapter controller) {
        Double lastProgress = null;
        String lastMessage = null;
        Object artifactValue = null;
        boolean sawArtifact = false;
        String terminalState = null;
        String terminalMessage = null;

        try {
            for (A2AEvent event : this) {
                if (event.kind() == A2AEvent.Kind.ARTIFACT) {
                    artifactValue = parseArtifactValue(event.artifactText());
                    sawArtifact = true;
                    continue;
                }
                // STATUS event
                if (event.progress() != null || event.message() != null) {
                    boolean changed = !sameOrNull(event.progress(), lastProgress)
                        || !equalsNullable(event.message(), lastMessage);
                    if (changed) {
                        double rawP = event.progress() != null
                            ? event.progress()
                            : (lastProgress != null ? lastProgress : 0.0d);
                        double clamped = Math.min(1.0d, Math.max(0.0d, rawP));
                        try {
                            controller.updateProgress(clamped, event.message());
                            if (event.progress() != null) {
                                lastProgress = event.progress();
                            }
                            lastMessage = event.message();
                        } catch (RuntimeException exc) {
                            // Do NOT advance lastProgress / lastMessage —
                            // the next event with the same delta still
                            // passes the equality check below and retries.
                            log.warn("A2AStream.bridge: controller.updateProgress failed (task={}, progress={}, msg={}) — will retry on next event",
                                taskId, event.progress(), event.message(), exc);
                        }
                    }
                }
                if (event.isFinal()) {
                    terminalState = event.state();
                    terminalMessage = event.message();
                    break;
                }
            }
        } finally {
            close();
        }

        if (terminalState != null) {
            String lower = terminalState.toLowerCase(java.util.Locale.ROOT);
            if (lower.equals("canceled") || lower.equals("cancelled")) {
                throw new A2AJobCanceledException(
                    terminalMessage != null ? terminalMessage : "A2A task " + taskId + " canceled");
            }
            if (lower.equals("failed")) {
                throw new A2AJobFailedException(
                    terminalMessage != null ? terminalMessage : "A2A task " + taskId + " failed");
            }
        }
        if (!sawArtifact) {
            // Stream closed without an artifact event AND without a
            // terminal failure — surface as failed so the user function
            // raises rather than silently returning null.
            throw new A2AJobFailedException(
                "A2A subscribe stream " + taskId + " ended without artifact");
        }
        return artifactValue;
    }

    @Override
    public void close() {
        if (closed) {
            return;
        }
        closed = true;
        try {
            // BufferedReader.close cascades to the underlying
            // InputStream which closes the HttpResponse body — releases
            // the JDK HttpClient connection back to its pool.
            reader.close();
        } catch (IOException exc) {
            log.debug("A2AStream.close: reader close raised (task={}): {}", taskId, exc.getMessage());
        }
    }

    private Object parseArtifactValue(String text) {
        if (text == null || text.isEmpty()) {
            return text == null ? "" : text;
        }
        try {
            JsonNode parsed = objectMapper.readTree(text);
            if (parsed == null || parsed.isNull()) {
                return text;
            }
            return objectMapper.convertValue(parsed, Object.class);
        } catch (Exception parseExc) {
            return text;
        }
    }

    private static boolean sameOrNull(Double a, Double b) {
        if (a == null && b == null) return true;
        if (a == null || b == null) return false;
        return a.doubleValue() == b.doubleValue();
    }

    private static boolean equalsNullable(String a, String b) {
        if (a == null && b == null) return true;
        if (a == null || b == null) return false;
        return a.equals(b);
    }

    /**
     * Pull-style SSE event iterator. Caches the next event so
     * {@link #hasNext} can do the I/O work needed to know if another
     * event is available without losing it for {@link #next}.
     */
    private final class SseIterator implements Iterator<A2AEvent> {

        private A2AEvent next;
        /** True once the terminal {@code isFinal=true} frame has been consumed. */
        private boolean terminated;

        @Override
        public boolean hasNext() {
            if (terminated) {
                return false;
            }
            if (next != null) {
                return true;
            }
            next = readNext();
            return next != null;
        }

        @Override
        public A2AEvent next() {
            if (!hasNext()) {
                throw new NoSuchElementException("A2AStream exhausted (task=" + taskId + ")");
            }
            A2AEvent ev = next;
            next = null;
            if (ev.isFinal()) {
                terminated = true;
                // Auto-close on the terminal frame so the connection
                // returns to the pool promptly without waiting on the
                // GC. close() is idempotent.
                close();
            }
            return ev;
        }

        /**
         * Read SSE lines until a complete event is parsed (returns
         * non-null) or the stream is exhausted (returns null). Skips
         * comment lines, keepalives, and unknown JSON-RPC envelope
         * shapes.
         */
        private A2AEvent readNext() {
            if (closed) {
                return null;
            }
            StringBuilder dataBuf = new StringBuilder();
            try {
                String line;
                while ((line = reader.readLine()) != null) {
                    if (line.isEmpty()) {
                        // Blank line — event boundary.
                        if (dataBuf.length() == 0) {
                            continue;
                        }
                        A2AEvent parsed = tryParse(dataBuf.toString());
                        dataBuf.setLength(0);
                        if (parsed != null) {
                            return parsed;
                        }
                        // Unknown envelope shape — skip and keep reading.
                        continue;
                    }
                    if (line.startsWith(":")) {
                        // SSE comment (keepalive) — ignore.
                        continue;
                    }
                    if (line.startsWith("data:")) {
                        // Strip "data:" prefix and one optional space; per
                        // SSE spec, multiple data: lines for the same event
                        // are joined with \n.
                        String payload = line.substring(5);
                        if (payload.startsWith(" ")) {
                            payload = payload.substring(1);
                        }
                        if (dataBuf.length() > 0) {
                            dataBuf.append('\n');
                        }
                        dataBuf.append(payload);
                        continue;
                    }
                    // event:/id:/retry:/unknown — ignore for v1.0.
                }

                // Stream ended. Flush any pending frame, then stop.
                if (dataBuf.length() > 0) {
                    return tryParse(dataBuf.toString());
                }
                return null;
            } catch (IOException exc) {
                // Transport error — close the stream so we don't leak
                // the connection, then surface as failed via the
                // bridging contract.
                close();
                throw new A2AException(
                    "A2AStream read failed for task " + taskId + ": " + exc.getMessage(), exc);
            }
        }

        private A2AEvent tryParse(String payload) {
            JsonNode envelope;
            try {
                envelope = objectMapper.readTree(payload);
            } catch (Exception exc) {
                log.debug("A2AStream: skipping non-JSON SSE frame (task={}): payload={}", taskId, payload);
                return null;
            }
            if (envelope == null || !envelope.isObject()) {
                return null;
            }
            JsonNode result = envelope.get("result");
            if (result == null || !result.isObject()) {
                return null;
            }

            // ARTIFACT events have the "artifact" key.
            JsonNode artifact = result.get("artifact");
            if (artifact != null && artifact.isObject()) {
                String text = extractFirstTextPart(artifact);
                if (text == null) {
                    text = "";
                }
                return new A2AEvent(A2AEvent.Kind.ARTIFACT, null, null, null, text, false, envelope);
            }

            // STATUS events have the "status" key.
            JsonNode status = result.get("status");
            if (status != null && status.isObject()) {
                String state = null;
                JsonNode stateNode = status.get("state");
                if (stateNode != null && !stateNode.isNull()) {
                    state = stateNode.asString();
                }
                String message = null;
                JsonNode msgObj = status.get("message");
                if (msgObj != null && msgObj.isObject()) {
                    message = extractFirstTextPart(msgObj);
                }
                Double progress = null;
                JsonNode metadata = result.get("metadata");
                if (metadata != null && metadata.isObject()) {
                    JsonNode pNode = metadata.get("progress");
                    if (pNode != null && !pNode.isNull()) {
                        if (pNode.isNumber()) {
                            progress = pNode.asDouble();
                        } else {
                            try {
                                progress = Double.parseDouble(pNode.asString());
                            } catch (Exception ignored) {
                                progress = null;
                            }
                        }
                    }
                }
                boolean isFinal = false;
                JsonNode finalNode = result.get("final");
                if (finalNode != null && !finalNode.isNull()) {
                    isFinal = finalNode.asBoolean(false);
                }
                return new A2AEvent(
                    A2AEvent.Kind.STATUS, state, progress, message, null, isFinal, envelope);
            }

            return null;
        }

        private String extractFirstTextPart(JsonNode container) {
            JsonNode parts = container.get("parts");
            if (!(parts instanceof ArrayNode array) || array.isEmpty()) {
                return null;
            }
            JsonNode first = array.get(0);
            if (first == null || !first.isObject()) {
                return null;
            }
            JsonNode text = first.get("text");
            if (text == null || text.isNull()) {
                return null;
            }
            return text.asString();
        }
    }
}

File: src/runtime/typescript/src/a2a/a2a-stream.ts
/**
 * Async iterator over parsed A2A SSE events — returned by
 * `A2AClient.subscribe` (issue #917).
 *
 * Implements `AsyncIterable<A2AEvent>` so callers can write
 * `for await (const event of stream) { ... }` idiomatically.
 * Implements TC39 `Symbol.asyncDispose` (Node 20+) for `await using`,
 * AND exposes `aclose()` as a manual fallback for older runtimes.
 *
 * Per A2A v1.0, client disconnect is a transient signal — the producer
 * continues running unless explicitly canceled. `bridge` therefore does
 * NOT POST `tasks/cancel` upstream when the mesh-side job is cancelled
 * (it just closes the SSE connection). Users who need cancel
 * propagation should use `submit` + `A2AJob.bridge` instead, which
 * polls for cancel between iterations and propagates upstream.
 *
 * Mirrors `mesh._a2a_consumer.A2AStream` (Python) and
 * `io.mcpmesh.a2a.A2AStream` (Java).
 */
import { JobController } from "@mcpmesh/core";
import type { A2AEvent } from "./a2a-event.js";
import {
  A2AJobCanceledError,
  A2AJobFailedError,
} from "./errors.js";
import { maybeJsonParse } from "./a2a-client.js";

const ASYNC_DISPOSE: typeof Symbol.asyncDispose | symbol =
  (Symbol as { asyncDispose?: symbol }).asyncDispose ??
  Symbol.for("Symbol.asyncDispose");

export class A2AStream implements AsyncIterable<A2AEvent> {
  private readonly response: Response;
  readonly taskId: string;
  private closed = false;
  private reader: ReadableStreamDefaultReader<Uint8Array> | null = null;

  constructor(response: Response, taskId: string) {
    this.response = response;
    this.taskId = taskId;
  }

  [Symbol.asyncIterator](): AsyncIterator<A2AEvent> {
    return this._iterate();
  }

  /**
   * TC39 explicit-resource-management hook — Node 20+ supports
   * `await using stream = await client.subscribe(...)`. The fallback
   * below uses Symbol.for so older runtimes still surface a usable
   * symbol on the prototype (idempotent close).
   */
  async [ASYNC_DISPOSE as typeof Symbol.asyncDispose](): Promise<void> {
    await this.aclose();
  }

  /** Mirror events into a JobController; return the final artifact value. */
  async bridge(controller: JobController): Promise<unknown> {
    if (controller == null) {
      throw new Error("A2AStream.bridge: controller must be non-null");
    }
    let lastProgress: number | undefined;
    let lastMessage: string | undefined;
    let artifactValue: unknown = "";
    let sawArtifact = false;
    let terminalState: string | undefined;
    let terminalMessage: string | undefined;

    try {
      for await (const event of this) {
        if (event.kind === "artifact") {
          const text = event.artifactText ?? "";
          artifactValue = !text ? text : maybeJsonParse(text);
          sawArtifact = true;
          continue;
        }
        if (event.progress !== undefined || event.message !== undefined) {
          const changed =
            event.progress !== lastProgress || event.message !== lastMessage;
          if (changed) {
            const rawP =
              event.progress !== undefined
                ? event.progress
                : lastProgress !== undefined
                  ? lastProgress
                  : 0.0;
            const clamped = Math.min(1.0, Math.max(0.0, rawP));
            try {
              await controller.updateProgress(clamped, event.message ?? null);
              if (event.progress !== undefined) lastProgress = event.progress;
              lastMessage = event.message;
            } catch (err) {
              // Do NOT advance lastProgress / lastMessage on delivery
              // failure — leaving them stale ensures the next event
              // still passes the equality check and retries.
              console.warn(
                `[a2a-stream] bridge: controller.updateProgress failed ` +
                  `(task=${this.taskId}, progress=${event.progress}, ` +
                  `msg=${event.message}) — will retry on next event: ` +
                  `${(err as Error)?.message ?? String(err)}`,
              );
            }
          }
        }
        if (event.final) {
          terminalState = event.state;
          terminalMessage = event.message;
          break;
        }
      }
    } finally {
      // Ensure the SSE stream is closed on any exit path (normal
      // completion, throw, etc.). aclose() is idempotent.
      await this.aclose();
    }

    if (terminalState) {
      const lower = terminalState.toLowerCase();
      if (lower === "canceled" || lower === "cancelled") {
        throw new A2AJobCanceledError(
          terminalMessage ?? `A2A task ${this.taskId} canceled`,
        );
      }
      if (lower === "failed") {
        throw new A2AJobFailedError(
          terminalMessage ?? `A2A task ${this.taskId} failed`,
        );
      }
      // Parity with A2AJob._terminalToArtifactOrThrow — completed
      // with no artifact event seen returns "" rather than throwing,
      // matching the polling-bridge semantics.
      if (lower === "completed" && !sawArtifact) {
        return "";
      }
    }
    if (!sawArtifact) {
      throw new A2AJobFailedError(
        `A2A subscribe stream ${this.taskId} ended without artifact`,
      );
    }
    return artifactValue;
  }

  /**
   * Explicit close — releases the reader lock + cancels the
   * underlying ReadableStream so the undici Agent reclaims the
   * connection promptly. Idempotent.
   */
  async aclose(): Promise<void> {
    if (this.closed) return;
    this.closed = true;
    try {
      if (this.reader) {
        try {
          await this.reader.cancel();
        } catch {
          // best-effort
        }
        try {
          this.reader.releaseLock();
        } catch {
          // best-effort
        }
        this.reader = null;
      } else if (this.response.body) {
        try {
          await this.response.body.cancel();
        } catch {
          // best-effort
        }
      }
    } catch {
      // swallow — close is best-effort
    }
  }

  private async *_iterate(): AsyncGenerator<A2AEvent> {
    if (this.closed) return;
    if (!this.response.body) {
      throw new A2AJobFailedError(
        `A2AStream(${this.taskId}): response body is null`,
      );
    }
    if (!this.reader) {
      this.reader = this.response.body.getReader();
    }
    const reader = this.reader;
    const decoder = new TextDecoder("utf-8");
    let pending = "";
    const dataBuf: string[] = [];

    const flushEvent = (): A2AEvent | null => {
      if (dataBuf.length === 0) return null;
      const payload = dataBuf.join("\n");
      dataBuf.length = 0;
      let envelope: unknown;
      try {
        envelope = JSON.parse(payload);
      } catch {
        // Skip non-JSON SSE frames silently — matches Python behavior.
        return null;
      }
      return parseSseEnvelope(envelope);
    };

    try {
      while (true) {
        const { value, done } = await reader.read();
        if (done) break;
        pending += decoder.decode(value, { stream: true });
        let nlIdx: number;
        while ((nlIdx = pending.search(/\r\n|\n/)) !== -1) {
          // Accept both LF and CRLF line endings.
          const isCrLf =
            pending.charCodeAt(nlIdx) === 13 &&
            pending.charCodeAt(nlIdx + 1) === 10;
          const line = pending.slice(0, nlIdx);
          pending = pending.slice(nlIdx + (isCrLf ? 2 : 1));
          if (line === "") {
            const event = flushEvent();
            if (event !== null) {
              yield event;
              if (event.final) {
                await this.aclose();
                return;
              }
            }
            continue;
          }
          if (line.startsWith(":")) {
            // SSE comment / keepalive — ignore.
            continue;
          }
          if (line.startsWith("data:")) {
            // Strip "data:" prefix and one optional space.
            let payload = line.slice(5);
            if (payload.startsWith(" ")) payload = payload.slice(1);
            dataBuf.push(payload);
            continue;
          }
          // event:/id:/retry:/unknown — ignore for v1.0.
        }
      }
      // Stream ended without a trailing blank — flush any pending frame.
      pending += decoder.decode();
      if (pending.length > 0 && pending.startsWith("data:")) {
        let payload = pending.slice(5);
        if (payload.startsWith(" ")) payload = payload.slice(1);
        dataBuf.push(payload);
      }
      const tail = flushEvent();
      if (tail !== null) {
        yield tail;
      }
    } finally {
      await this.aclose();
    }
  }
}

function parseSseEnvelope(envelope: unknown): A2AEvent | null {
  if (!envelope || typeof envelope !== "object") return null;
  const result = (envelope as Record<string, unknown>).result;
  if (!result || typeof result !== "object") return null;
  const r = result as Record<string, unknown>;

  // ARTIFACT events have the "artifact" key.
  const artifact = r.artifact;
  if (artifact && typeof artifact === "object") {
    const text = extractFirstTextPart(artifact as Record<string, unknown>) ?? "";
    return {
      kind: "artifact",
      artifactText: text,
      final: false,
      raw: envelope,
    };
  }

  // STATUS events have the "status" key.
  const status = r.status;
  if (status && typeof status === "object") {
    const statusObj = status as Record<string, unknown>;
    const stateRaw = statusObj.state;
    const state = typeof stateRaw === "string" ? stateRaw : undefined;
    const msgObj = statusObj.message;
    let message: string | undefined;
    if (msgObj && typeof msgObj === "object") {
      message = extractFirstTextPart(msgObj as Record<string, unknown>);
    }
    let progress: number | undefined;
    const metadata = r.metadata;
    if (metadata && typeof metadata === "object") {
      const p = (metadata as Record<string, unknown>).progress;
      if (typeof p === "number" && Number.isFinite(p)) progress = p;
      else if (typeof p === "string") {
        const n = Number.parseFloat(p);
        if (Number.isFinite(n)) progress = n;
      }
    }
    const finalRaw = r.final;
    const final = finalRaw === true;
    return {
      kind: "status",
      state,
      progress,
      message,
      final,
      raw: envelope,
    };
  }

  return null;
}

function extractFirstTextPart(
  container: Record<string, unknown>,
): string | undefined {
  const parts = container.parts;
  if (!Array.isArray(parts) || parts.length === 0) return undefined;
  const first = parts[0];
  if (!first || typeof first !== "object") return undefined;
  const text = (first as Record<string, unknown>).text;
  return typeof text === "string" ? text : undefined;
}
